src/evaluators/DoWhileStatement.js

```diff
@@ -33,34 +34,67 @@ export default function(
+  let resultOrDiagnostic = realm.evaluateWithUndoForDiagnostic(() => {
+    while (true) {
+      // a. Let stmt be the result of evaluating Statement.
+      let stmt = env.evaluateCompletion(body, strictCode);
+      //todo: check if stmt is a PossiblyNormalCompletion and defer to fixpoint computation below
+      invariant(stmt instanceof Value || stmt instanceof AbruptCompletion);
 
-    // b. If LoopContinues(stmt, labelSet) is false, return Completion(UpdateEmpty(stmt, V)).
-    if (LoopContinues(realm, stmt, labelSet) === false) {
-      invariant(stmt instanceof AbruptCompletion);
-      // ECMA262 13.1.7
-      if (stmt instanceof BreakCompletion) {
-        if (!stmt.target) return (UpdateEmpty(realm, stmt, V): any).value;
+      // b. If LoopContinues(stmt, labelSet) is false, return Completion(UpdateEmpty(stmt, V)).
+      if (LoopContinues(realm, stmt, labelSet) === false) {
+        invariant(stmt instanceof AbruptCompletion);
+        // ECMA262 13.1.7
+        if (stmt instanceof BreakCompletion) {
+          if (!stmt.target) return (UpdateEmpty(realm, stmt, V): any).value;
+        }
+        throw UpdateEmpty(realm, stmt, V);
       }
-      throw UpdateEmpty(realm, stmt, V);
-    }
 
-    // c. If stmt.[[Value]] is not empty, let V be stmt.[[Value]].
-    let resultValue = InternalGetResultValue(realm, stmt);
-    if (!(resultValue instanceof EmptyValue)) V = resultValue;
+      // c. If stmt.[[Value]] is not empty, let V be stmt.[[Value]].
+      let resultValue = InternalGetResultValue(realm, stmt);
+      if (!(resultValue instanceof EmptyValue)) V = resultValue;
 
-    // d. Let exprRef be the result of evaluating Expression.
-    let exprRef = env.evaluate(test, strictCode);
+      // d. Let exprRef be the result of evaluating Expression.
+      let exprRef = env.evaluate(test, strictCode);
 
-    // e. Let exprValue be ? GetValue(exprRef).
-    let exprValue = Environment.GetValue(realm, exprRef);
+      // e. Let exprValue be ? GetValue(exprRef).
+      let exprValue = Environment.GetConditionValue(realm, exprRef);
+
+      // f. If ToBoolean(exprValue) is false, return NormalCompletion(V).
+      if (To.ToBooleanPartial(realm, exprValue) === false) return V;
+    }
+    invariant(false);
+  });
+  if (resultOrDiagnostic instanceof Value) return resultOrDiagnostic;
 
-    // f. If ToBoolean(exprValue) is false, return NormalCompletion(V).
-    if (To.ToBooleanPartial(realm, exprValue) === false) return V;
+  // If we get here then unrolling the loop did not work, possibly because the value of the loop condition is not known,
+  // so instead try to compute a fixpoint for it
+  let ftest = () => {
+    let exprRef = env.evaluate(test, strictCode);
+    return Environment.GetConditionValue(realm, exprRef);
+  };
+  let fbody = () => env.evaluateCompletion(body, strictCode);
+  let effects = realm.evaluateForFixpointEffects(ftest, fbody);
+  if (effects !== undefined) {
+    let [outsideEffects, insideEffects] = effects;
```

Comment:
What is insideEffects here?

Reply:
The effects that summarizes the state changes of an arbitrary iteration of the loop body. That is used to generate code for the loop body. The outsideEffects is used to summary the entire loop for the subsequent code.

Reply:
Maybe worth documenting these two effects in evaluateForFixpointEffects().